framework/src/modules/interoperability/base_cross_chain_update_command.ts

```diff
@@ -174,20 +174,131 @@ export abstract class BaseCrossChainUpdateCommand<
+					);
+					await method.beforeCrossChainCommandExecute(context);
+				}
+			}
+		} catch (error) {
+			// revert state to baseSnapshot
+			context.eventQueue.restoreSnapshot(baseEventSnapshotID);
+			context.stateStore.restoreSnapshot(baseStateSnapshotID);
+
+			logger.info(
+				{ err: error as Error, moduleName: ccm.module, commandName: ccm.crossChainCommand },
+				'Fail to execute beforeCrossChainCommandExecute.',
+			);
+
+			await this.internalMethod.terminateChainInternal(context, ccm.sendingChainID);
+			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
+				ccm,
+				result: CCMProcessedResult.DISCARDED,
+				code: CCMProcessedCode.INVALID_CCM_BEFORE_CCC_EXECUTION_EXCEPTION,
+			});
+			return false;
+		}
+
+		return true;
+	}
+
+	private async _afterCrossChainCommandExecute(
+		context: CrossChainMessageContext,
+		baseEventSnapshotID: number,
+		baseStateSnapshotID: number,
+	): Promise<boolean> {
+		const { ccm, logger } = context;
+		const { ccmID } = getEncodedCCMAndID(ccm);
+
+		try {
+			// Call the beforeCrossChainCommandExecute functions from other modules.
+			// For example, the Token module assigns the message fee to the CCU sender.
+			for (const [module, method] of this.interoperableCCMethods.entries()) {
+				if (method.afterCrossChainCommandExecute) {
+					logger.debug(
+						{
+							moduleName: module,
+							commandName: ccm.crossChainCommand,
+							ccmID: ccmID.toString('hex'),
+						},
+						'Execute afterCrossChainCommandExecute',
+					);
+					await method.afterCrossChainCommandExecute(context);
+				}
+			}
+		} catch (error) {
+			// revert state to baseSnapshot
+			context.eventQueue.restoreSnapshot(baseEventSnapshotID);
+			context.stateStore.restoreSnapshot(baseStateSnapshotID);
+
+			logger.info(
+				{ err: error as Error, moduleName: ccm.module, commandName: ccm.crossChainCommand },
+				'Fail to execute afterCrossChainCommandExecute.',
+			);
+
+			await this.internalMethod.terminateChainInternal(context, ccm.sendingChainID);
+			this.events.get(CcmProcessedEvent).log(context, ccm.sendingChainID, ccm.receivingChainID, {
+				ccm,
+				result: CCMProcessedResult.DISCARDED,
+				code: CCMProcessedCode.INVALID_CCM_AFTER_CCC_EXECUTION_EXCEPTION,
+			});
+			return false;
+		}
+
+		return true;
+	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0049.md#apply
 	protected async apply(context: CrossChainMessageContext): Promise<void> {
 		const { ccm, logger } = context;
 		const { ccmID, encodedCCM } = getEncodedCCMAndID(ccm);
 		const valid = await this.verifyCCM(context, ccmID);
 		if (!valid) {
 			return;
 		}
-		const commands = this.ccCommands.get(ccm.module);
-		if (!commands) {
+
+		// Create a state snapshot.
+		const baseEventSnapshotID = context.eventQueue.createSnapshot();
+		const baseStateSnapshotID = context.stateStore.createSnapshot();
+
+		const crossChainCommands = this.ccCommands.get(ccm.module);
```

Comment:
I'm a bit confused by this interface, basically here we check if the module is supported?